controlplane/kubeadm/internal/workload_cluster.go

```diff
@@ -274,6 +291,151 @@ func (w *Workload) UpdateKubernetesVersionInKubeadmConfigMap(ctx context.Context
+	}
+
+	updatedCorefile, err := w.Migrator.Migrate(currentCoreDNSVersion, toCoreDNSVersion, currentCoreDNSCorefile, false)
+	if err != nil {
+		return errors.Wrap(err, "unable to migrate CoreDNS corefile")
+	}
+	if err := w.Client.Update(ctx, &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "coredns",
+			Namespace: metav1.NamespaceSystem,
+		},
+		Data: map[string]string{
+			"Corefile":        updatedCorefile,
+			"Corefile-backup": currentCoreDNSCorefile,
+		},
+	}); err != nil {
+		return errors.Wrap(err, "unable to update CoreDNS config map")
+	}
+
+	return nil
+}
+
+// UpdateCoreDNSDeployment will patch the deployment image to the
+// imageRepo:imageTag in the KCP dns. It will also ensure the volume of the
+// deployment uses the Corefile key of the coredns configmap.
+func (w *Workload) UpdateCoreDNSDeployment(ctx context.Context, dns *kubeadmv1.DNS) error {
+	image := fmt.Sprintf("%s:%s", dns.ImageRepository, dns.ImageTag)
+	patchJSON := fmt.Sprintf(
+		`{"spec":{"template":{"spec":{"containers":[{"name": "coredns", "image":"%s"}],"volumes":[{"name": "config-volume", "configMap":{"name": "coredns", "items":[{"key": "%[2]s", "path": "%[2]s"}]}}]}}}}`,
+		image,
+		corefileKey,
+	)
+	w.patchCoreDNSDeployment(ctx, patchJSON)
+
+	return nil
+}
+
+func (w *Workload) patchCoreDNSDeployment(ctx context.Context, patchJSON string) error {
+	corednsKey := ctrlclient.ObjectKey{Name: coreDNSKey, Namespace: metav1.NamespaceSystem}
+	orig := &appsv1.Deployment{}
+	if err := w.Client.Get(ctx, corednsKey, orig); err != nil {
+		return errors.Wrapf(err, "unable to get %s deployment from target cluster", corednsKey.String())
+	}
+	patch := ctrlclient.RawPatch(types.StrategicMergePatchType, []byte(patchJSON))
+
+	if err := w.Client.Patch(ctx, orig, patch); err != nil {
+		return errors.Wrapf(err, "unable to patch %s deployment in target cluster", corednsKey.String())
+	}
+	return nil
+}
+
+// GetCoreDNSInfo gets the current CoreDNS installed and the current Corefile Configuration of CoreDNS.
+func (w *Workload) GetCoreDNSInfo(ctx context.Context) (string, string, error) {
+	corednsKey := ctrlclient.ObjectKey{Name: coreDNSKey, Namespace: metav1.NamespaceSystem}
+	cm := &corev1.ConfigMap{}
+	if err := w.Client.Get(ctx, corednsKey, cm); err != nil {
+		return "", "", errors.Wrapf(err, "error getting %s config map from target cluster", corednsKey.String())
+	}
+	corefile, ok := cm.Data[corefileKey]
+	if !ok {
+		return "", "", errors.New("unable to find the CoreDNS Corefile data")
+	}
+
+	deployment := &appsv1.Deployment{}
+	if err := w.Client.Get(ctx, corednsKey, deployment); err != nil {
+		return "", "", errors.Wrapf(err, "error getting %s deployment from target cluster", corednsKey.String())
+	}
+	addonImage := deployment.Spec.Template.Spec.Containers[0].Image
+	addonImageParts := strings.Split(addonImage, ":")
+	addonVersion := addonImageParts[len(addonImageParts)-1]
```

Comment:
We should probably use the `distribution/reference` library like we're doing in other places:
- https://github.com/kubernetes-sigs/cluster-api/blob/b51a61ec6b1007347524fb65766a8b351e4de0bb/cmd/clusterctl/client/config/imagemeta_client.go
- https://github.com/kubernetes-sigs/cluster-api/pull/2559/files#diff-babaf075698523dd601d02d21570c460R31